### PR TITLE
Moved to datetime

### PR DESCRIPTION
**Pull request by anurag-ks**
Task - https://www.google-melange.com/gci/task/view/google/gci2014/5890689128726528
Todo - https://github.com/edengci/eden/blob/master/modules/s3db/budget.py#L1302
This PR also contains the fixups for the mistakes pointed out by @flavour

### CODE REVIEW

**Comment by ptressel** [comment]
Anurag -- the "Fixups" commit is good, but shouldn't be in this PR -- it needs to go with the combined PR to trunk, where the commits it is fixing are -- the staging_done branch of edengci/eden.  (We don't have a procedure for updating that yet, but maybe we can use your commit as a test case.)  So...can you move the Fixups commit over to a different branch, that is based on staging_done?  Here's one way to do that, which is much like the little ritual we did to fix up this PR, e.g.:

git fetch edengci staging_done
git checkout -b staging_done_fixups staging_done
git cherry-pick cc42b5084a1ba8567b685bd178535786ef49c864
git status

If the cherry-pick gets merge conflicts, then it will stop before doing the commit, and you'll have to fix the conflicts and do the commit.  If you included all your work from staging_done in the Fixups commit, then it will _definitely_ get merge conflicts.  Look at the cherry-pick help info to see what it does on a conflict:
git help cherry-pick

When you fix the conflicts, you should end up with the same final form of your changes, after your Fixups commit, except that there may be other students' work from staging_done, in those same files.

Then test your changes, and when you're satisfied, push staging_done_fixups to GitHub, so we can go over it before transferring it to the edengci/eden staging_done.

After you're sure your Fixups commit is safely transferred and committed, then you can remove it from this branch and push again, so this has just the one new commit.

**Comment by anurag-ks** [comment]
@ptressel - https://github.com/edengci/eden/pull/73

File: modules/s3db/msg.py [hunk 1]
```diff
@@ -582,7 +582,8 @@ def model(self):
         tablename = "msg_email_channel"
         define_table(tablename,
                      # Instance
-                     super_link("channel_id", "msg_channel"),
+                     super_link("message_id", "msg_message"),
+                     self.msg_channel_id(),
```
Comment (flavour) [inline modules/s3db/msg.py]: err, no.
This is NOT an instance of msg_message.
As an instance of msg_channel it doesn't need the self.msg_channel_id(), but rather should keep what ti is.
self.msg_channel_id is for components of channel

File: modules/s3db/budget.py [hunk 2]
```diff
@@ -1299,13 +1299,12 @@ def model(self):
...
                                           represent = self.budget_CostItemRepresent(),
                                           ),
-                          # @ToDo: s3_datetime
-                          s3_date("start_date",
-                                  label = T("Start Date")
-                                  ),
-                          s3_date("end_date",
-                                  label = T("End Date")
-                                  ),
+                          s3_datetime("start_date",
```
Comment (flavour) [inline modules/s3db/budget.py:1302]: I would be very wary of accepting this as-is.
As-is this is obviously trivial....so why was it left as a ToDo?
Have you been through the module & tested that all still works with datetimes?
- ideally writing some Tests could help with proving this.

UX-wise, I would think that we would want to enter dates in most cases as is done now, rather than a datetime widget (look for the s3_datetime option to allow this)

There is also potentially a data migration issue...I'm not personally familiar with the issues when moving dates to datetimes. I'm not too worried about this personally as I'm not aware of the module being used in Production yet, however in general it is somethign to be considered when making such changes.

File: controllers/deploy.py [hunk 3]
```diff
@@ -637,8 +637,6 @@ def postp(r, output):
 def email_channel():
     """
         RESTful CRUD controller for Inbound Email channels
-
-        @ToDo: Allow selection of a specific Channel for Alerts
```
Comment (flavour) [inline controllers/deploy.py:641]: Why is this ToDo removed?
Has it been actioned? Where?

Comment (ptressel) [inline controllers/deploy.py:641]: Was this an accidental change?  It doesn't seem related to datetimes.

Comment (anurag-ks) [inline controllers/deploy.py:641]: Yeah sorry for that, i was testing out some stuff (I might i have forgotten to change my branch to local), I'll revert them.

**Comment by ptressel** [comment]
Need to know which ToDo this is doing.  Since that ToDo should be removed when done, you can point to it on GitHub by linking to the file in a specific commit that contains the ToDo.

File: modules/s3db/budget.py (hunk 2 above)
Comment (anurag-ks) [inline modules/s3db/budget.py:1302]: @flavour - https://github.com/edengci/eden/commit/a264c2b03751d12cbecc09bdbb53275b5b63dc54 This is when you added the todo ;)

Comment (flavour) [inline modules/s3db/budget.py:1302]: That doesn't explain why it wasn't done right away.
As I said, & Dominic has said before, if it was as trivial as you've done then it would have been done right away.
A ToDo is placed because it requires more work than is currently avaialble to evaluate all the implications and come up with a full plan.

**Comment by anurag-ks** [comment]
While testing it out, i am getting this error - http://paste.ubuntu.com/9665904/ I tried changing the represent but still no luck, so I think I must leave this for now.
